ci(aio): change AIO preview server stuff to use CircleCI

petebacondarwin opened the pull request:
The system now waits for a CircleCI webhook to tell the upload server that a build has completed.
The upload server then pulls down the build artifact and passes it along as before to the rest of the preview creation logic.

This means that there is no need to store a secret on CircleCI for pushing the artifacts to the upload server; which means that we no longer rely upon JWT.

The downloaded artifact is stored in a temp location; this is cleared out along with the actual preview builds by the periodic cleanup task.

There are some (self-serving) changes to the testing of components that were not significantly changed for this PR. This includes more use of `async/await` for managing async tests; and also making some API public (rather than protected). This allows us to unit test, while maintaining type safety. Previously, we had to cast to `any` which meant that changes to the API may not be passed through to the unit tests.

TODO:

- [x] check for CI success before processing
- [x] consider not downloading an artifact if it has been downloaded already
- [x] fix up / write more e2e tests
- [x] update the docs
- [x] consider renaming "upload server" to "preview server"
- [x] configure the rest of the environment variables for the new processing
- [x] ascertain whether we need a CircleCI access token for making the GET requests
- [ ] work out how (and when) to run the PWA-score test; probably by the preview-server triggering a specific CircleCI job via API

IgorMinar commented on aio/scripts/build-artifacts.sh:
```diff
@@ -0,0 +1,24 @@
+source "`dirname $0`/../../scripts/ci/env.sh" print
+
+readonly INPUT_DIR=dist/
+readonly OUTPUT_FILE=tmp/snapshot.tar.gz
```
can we call this file with a more specific name? maybe `aio-snapshot.tgz`?

IgorMinar reviewed (commented):
at a high level this looks good. I'll let @gkalpak take a more thorough look through all the changes.

petebacondarwin committed:
ci(docs-infra): update upload-server to run on node.js v10

petebacondarwin committed:
ci(docs-infra): move AIO preview deployment to CircleCI

Now instead of pushing the AIO build artifacts to the preview server
from inside a Travis job, the artifacts are built and hosted on the
CircleCI infrastructure. The preview server will then pull these
down after being triggered by a CircleCI build webhook.

petebacondarwin committed:
docs(docs-infra): update the preview server documentation

petebacondarwin committed:
ci(docs-infra): add helper scripts for running TDD in Docker

gkalpak reviewed (commented):
Half-way through (mostly reviewed anything outside `scripts-js/`).

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/Dockerfile:
```diff
@@ -27,34 +31,39 @@ ARG      AIO_NGINX_PORT_HTTP=80
 ARG TEST_AIO_UPLOAD_HOSTNAME=upload.localhost
 ARG      AIO_UPLOAD_MAX_SIZE=20971520
-ARG TEST_AIO_UPLOAD_MAX_SIZE=20971520
+ARG TEST_AIO_UPLOAD_MAX_SIZE=200
```
This is because we want to test failures where the artifact is too large. Rather than having to create an enormous file, I just set this a "bit" lower :-)

gkalpak replied:
But why set it a "bit" lower (for tests)?

petebacondarwin replied:
I don't understand the question. I set it lower because in the tests we now actually create a mock file that is returned from the mock CircleCI API. To test that the max-size check is working we would have to create a mock file that is larger than 20971520 bytes. What is the point of that when we can just set the limit to something smaller?

petebacondarwin commented on scripts/ci/env.sh:
```diff
@@ -135,4 +135,35 @@ if [[ ${print} == "print" ]]; then
+
+  # Extract commit range (or single commit)
+  export CI_COMMIT_RANGE=$(echo "${CIRCLE_COMPARE_URL}" | cut -d/ -f7)
+  # Fix single commit, unfortunately we don't always get a commit range from Circle CI
```
It should be implied by this comment. For the purposes of this PR, this change is no longer needed as I moved the commit checking to the preview server.
I might revert this stuff.

petebacondarwin replied:
Yes I think we can revert this whole commit

petebacondarwin commented on scripts/ci/env.sh:
```diff
@@ -135,4 +135,35 @@ if [[ ${print} == "print" ]]; then
+  export CI_COMMIT_RANGE=$(echo "${CIRCLE_COMPARE_URL}" | cut -d/ -f7)
+  # Fix single commit, unfortunately we don't always get a commit range from Circle CI
+  if [[ $CI_COMMIT_RANGE != *"..."* ]]; then
+    export CI_COMMIT_RANGE="${CI_COMMIT}...${CI_COMMIT}"
```
+1

petebacondarwin commented on aio/scripts/deploy-preview.sh:
```diff
@@ -1,58 +0,0 @@
-
-  # Check the bundle sizes.
-  yarn payload-size
-)
```
+1

petebacondarwin commented on aio/scripts/build-artifacts.sh:
```diff
@@ -0,0 +1,27 @@
+  tar --create --gzip --directory "$INPUT_DIR" --file "$OUTPUT_FILE" .
+
+  # TODO(pete): work out how (and when) to run the PWA-score test
+  # probably by the preview-server triggering a specific CircleCI job via API
```
Done: https://github.com/angular/angular/issues/23818

petebacondarwin commented on aio/scripts/build-artifacts.sh:
```diff
@@ -0,0 +1,27 @@
+  # probably by the preview-server triggering a specific CircleCI job via API
+
+  # Check the bundle sizes.
+  yarn payload-size
```
Moved.

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/scripts-sh/verify-setup.sh:
```diff
@@ -31,10 +32,13 @@ function onExit {
 
 # Give the upload-server some time to start :(
-countdown "Starting" 5 > /dev/fd/3
+countdown "Starting" 2 > /dev/fd/3
```
It was just taking sooooo long... I'll revert now :-)

petebacondarwin commented on aio/aio-builds-setup/docs/overview--general.md:
```diff
@@ -32,34 +33,24 @@ This section gives a brief summary of the several operations performed on CI and
 More info on how to set things up on CI can be found [here](misc--integrate-with-ci.md).
 
 
 ### Uploading build artifacts
```
Right but then we need to agree to rename `upload-server` to something else :-P

petebacondarwin commented on aio/aio-builds-setup/docs/vm-setup--set-up-secrets.md:
```diff
@@ -28,25 +29,15 @@ not clear if/how they can be used in tests.
-
+2. `CIRCLECI_TOKEN`
+   - Visit https://circleci.com/gh/angular/angular/edit#api
+   - Create an API token with `Build Artifacts` scope
```
I believe hitting the general API for such info doesn't even require authentication!

petebacondarwin replied:
Honestly I think readonly access to download the artifacts doesn't even need authentication!

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/scripts-sh/test-env.sh:
```diff
@@ -0,0 +1,19 @@
+  # Set up env variables for testing
```
Done

gkalpak replied:
Are you sure?

petebacondarwin replied:
They are actually documented in [misc--debug-docker-container.md](https://github.com/petebacondarwin/angular/blob/e05aeae020d37202367b3bea6f059b215a8b3304/aio/aio-builds-setup/docs/misc--debug-docker-container.md). I'll add a bit to `overview--scripts-and-commands.md` for good measure.

petebacondarwin committed:
ci(docs-infra): move the payload-size check to the test job

petebacondarwin committed:
ci(docs-infra): factor out the aio-builds-setup environment variables

petebacondarwin committed:
ci(docs-infra): change AIO preview server stuff to pull builds from CircleCI

Previously, Travis pushed the build artitfacts to the preview server.
This required us to use JWT to secure the POST request from Travis, to
ensure we couldn't receive malicious builds.

JWT has been deprecated and we are moving our builds to CircleCI.

This commit rewrites the TypeScript part of the preview server that
handles converting build artifact into hosted previews of the docs.

petebacondarwin commented:
@gkalpak I have addressed yourr comments and reordered the commits. But I still have some tidying up to do in the TypeScript files - later today.

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/scripts-js/lib/common/github-api.ts:
```diff
@@ -28,29 +28,17 @@ export class GithubApi {
-  }
-
-  protected getPaginated<T>(pathname: string, baseParams: RequestParams = {}, currentPage: number = 0): Promise<T[]> {
+  public getPaginated<T>(pathname: string, baseParams: RequestParams = {}, currentPage: number = 0): Promise<T[]> {
```
Note that the specific GitHub API classes (like Teams and Pull Requests) now use this helper class by composition rather than inheritance. Therefore I had to make `getPaginated` a public method to expose it to these classes.

gkalpak replied:
Why using composition over inheritance?

petebacondarwin replied:
Composition provides a number of benefits. We only need to create on instance of the underlying GithubApi service, we can test that service independently of the subclasses that use it; and the clients of this service get a clean public API to work against.

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/scripts-js/lib/common/utils.ts:
```diff
@@ -1,16 +1,73 @@
+    } catch (error) {
+      console.error(error.stack);
+      process.exit(1);
+    }
```
Throwing and catching is simply to capture the stack trace to make debugging easier.

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/scripts-js/lib/upload-server/index-preverify-pr.ts:
```diff
@@ -7,16 +10,17 @@ _main();
+  const githubApi = new GithubApi(githubToken);
+  const prs = new GithubPullRequests(githubApi, githubOrg, githubRepo);
+  const teams = new GithubTeams(githubApi, githubOrg);
+  const buildVerifier = new BuildVerifier(prs, teams, allowedTeamSlugs, trustedPrLabel);
```
I wonder if we need to use a Dependency Injection framework here? :-P

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/scripts-js/lib/upload-server/upload-server-factory.ts:
```diff
@@ -2,49 +2,49 @@
 
 // Classes
-class UploadServerFactory {
+export class UploadServerFactory {
```
I converted these methods to `static` because really creating an instance of this class was unnecessary - there is not state in the class. The only reason I didn't pull them out into independent exported functions is that it is nice to have them in a container so that it is simple to spy/mock them for testing.

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/scripts-js/lib/upload-server/upload-server-factory.ts:
```diff
@@ -55,17 +55,87 @@ class UploadServerFactory {
+    // RESPOND TO IS-ALIVE PING
+    middleware.get(/^\/health-check\/?$/, (_req, res) => res.sendStatus(200));
+
+    // CIRCLE_CI BUILD COMPLETE WEBHOOK
```
Here is the meat of the change to this file - the new handler for the `circle-build` request.

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/scripts-js/lib/upload-server/upload-server-factory.ts:
```diff
@@ -55,17 +55,87 @@ class UploadServerFactory {
+      }
+    });
+
+    // GITHUB PR UPDATED WEBHOOK
```
I don't believe I made any substantive change to the `pr-updated` request handler, other than a sprinkling of `await`.

mary-poppins commented:
You can preview 2c4fc68 at https://pr23576-2c4fc68.ngbuilds.io/.

mary-poppins commented:
You can preview aa320f3 at https://pr23576-aa320f3.ngbuilds.io/.

petebacondarwin committed:
ci(docs-infra): improve preview-server logging

mary-poppins commented:
You can preview 88c3d95 at https://pr23576-88c3d95.ngbuilds.io/.

mary-poppins commented:
You can preview ad4f809 at https://pr23576-ad4f809.ngbuilds.io/.

mary-poppins commented:
You can preview 2c4fc68 at https://pr23576-2c4fc68.ngbuilds.io/.

mary-poppins commented:
You can preview 1fe1da4 at https://pr23576-1fe1da4.ngbuilds.io/.

mary-poppins commented:
You can preview e05aeae at https://pr23576-e05aeae.ngbuilds.io/.

gkalpak commented on .circleci/config.yml:
```diff
@@ -112,6 +112,22 @@ jobs:
+      - store_artifacts:
+          # The `path` needs to be kept in synch with the value of
+          # `OUTPUT_FILE` in `aio/scripts/build-artifacts.sh`
+          path: aio/tmp/snapshot.tgz
```
Would env vars be supported here? If so we could have the values in one place only.

petebacondarwin replied:
Good idea. I'll look into that...

petebacondarwin replied:
Nice!
This works.
https://circleci.com/gh/angular/angular/43101

gkalpak commented on aio/aio-builds-setup/dockerbuild/Dockerfile:
```diff
@@ -27,34 +34,36 @@ ARG      AIO_NGINX_PORT_HTTP=80
+    AIO_UPLOAD_HOSTNAME=$AIO_UPLOAD_HOSTNAME                      TEST_AIO_UPLOAD_HOSTNAME=$TEST_AIO_UPLOAD_HOSTNAME                      \
+    AIO_UPLOAD_MAX_SIZE=$AIO_UPLOAD_MAX_SIZE                      TEST_AIO_UPLOAD_MAX_SIZE=$TEST_AIO_UPLOAD_MAX_SIZE                      \
+    AIO_UPLOAD_PORT=$AIO_UPLOAD_PORT                              TEST_AIO_UPLOAD_PORT=$TEST_AIO_UPLOAD_PORT                              \
+    AIO_ARTIFACT_PATH=$AIO_ARTIFACT_PATH                          TEST_AIO_ARTIFACT_PATH=$TEST_AIO_ARTIFACT_PATH                          \
```
Not alphabetic order :sob:

gkalpak commented on aio/aio-builds-setup/docs/_TOC.md:
```diff
@@ -7,7 +7,6 @@
 - [Available scripts and commands](overview--scripts-and-commands.md)
 - [HTTP status codes](overview--http-status-codes.md)
 
-
```
Still no good 🙅‍♂️

petebacondarwin replied:
:-(

gkalpak commented on aio/scripts/build-artifacts.sh:
```diff
@@ -0,0 +1,19 @@
+#!/usr/bin/env bash
+
+set -eu -o pipefail
```
You can(/should) use `-x` now.

gkalpak commented on aio/aio-builds-setup/docs/image-config--environment-variables.md:
```diff
@@ -55,3 +60,8 @@ you don't need to specify values for those.
   The port number on which the Node.js upload-server listens for HTTP connections. This is used by
   nginx for delegating upload requests and also for performing a periodic health-check.
+
+- `AIO_ARTIFACT_PATH`:
```
Not a firm believer in alphabetic order, ha? 😞

gkalpak commented on aio/aio-builds-setup/dockerbuild/scripts-sh/run-tests.sh:
```diff
@@ -0,0 +1,3 @@
+rm /var/log/aio/*
```
Still feeling uncomfortable. 
As discussed, we could have these scripts in a directory that is not included in the docker container by default, but could be explicitly attached (e.g. as a volume).

petebacondarwin replied:
This file no longer exists. It has been replaced by `verify-setup-and-log`, which no longer deletes anything:

```
aio-verify-setup
ls -t /var/log/aio/upload-server-verify* | head -1 | xargs cat
```

gkalpak commented on aio/aio-builds-setup/dockerbuild/scripts-js/lib/common/env-variables.ts:
```diff
@@ -0,0 +1,19 @@
+import {getEnvVar} from './utils';
+
+export const AIO_ARTIFACT_PATH = getEnvVar('AIO_ARTIFACT_PATH');
```
Alphabetic order 🎊 ❤️ 💯

gkalpak commented on aio/aio-builds-setup/dockerbuild/scripts-js/lib/common/constants.ts:
```diff
@@ -1,3 +1,4 @@
 // Constants
 export const HIDDEN_DIR_PREFIX = 'hidden--';
 export const SHORT_SHA_LEN = 7;
+export const AIO_DOWNLOADS_DIR = '/tmp/aio-downloads';
```
No alphabetic order 🙄 😛

gkalpak commented on aio/aio-builds-setup/dockerbuild/scripts-js/test/clean-up/build-cleaner.spec.ts:
```diff
@@ -1,143 +1,184 @@
+    });
+
+
+    it('should throw if \'githubOrg\' is empty', () => {
```
`githubOrg` --> `githubRepo`

gkalpak commented on aio/aio-builds-setup/dockerbuild/scripts-js/lib/common/utils.ts:
```diff
@@ -1,17 +1,84 @@
+ * Get an environment variable.
+ * @param name The name of the environment variable.
+ * @param isOptional True if the variable is optional.
+ * @returns The value of the variable or "" if it is optional.
```
_if it is optional_ --> _if it is optional and undefined_ :grin:

gkalpak commented on aio/aio-builds-setup/dockerbuild/scripts-js/lib/common/github-api.ts:
```diff
@@ -67,7 +55,19 @@ export class GithubApi {
+    return `${pathname}${joiner}${search}`;
+  }
+
+  protected request<T>(method: string, path: string, data: any = null) {
```
Not a firm believer in return type annotations, ha? 😛

petebacondarwin replied:
IMO types in TypeScript should be added where necessary - i.e. where the compiler cannot imply them. Otherwise you are adding unnecessary noise to the code and adding to your maintenance effort in the long run.

<sub>(Having to declare types everywhere is one of the reasons I got out of .NET/Java development)</sub>

IgorMinar replied:
for what it's worth, I think it's useful to explicitly define the return value for any significant api. it just makes reading the code so much easier and faster without having to read the implementation.

gkalpak commented on aio/aio-builds-setup/dockerbuild/scripts-js/lib/upload-server/upload-server-factory.ts:
```diff
@@ -2,70 +2,159 @@
+        }
+
+        assert(cfg.githubOrg === org, `Invalid webhook: expected "${org}" to equal "${cfg.githubOrg}".`);
+        assert(cfg.githubRepo === repo, `Invalid webhook: expected "${repo}" to equal "${cfg.githubRepo}".`);
```
I would mention the field that failed, because it is not always clear (e.g. both `org` and `repo` is `angular`).

gkalpak commented on aio/aio-builds-setup/dockerbuild/scripts-js/lib/upload-server/upload-server-factory.ts:
```diff
@@ -2,70 +2,159 @@
+        if (!await buildVerifier.getSignificantFilesChanged(pr, new RegExp(cfg.significantFilesPattern))) {
+          logger.log(`PR:${pr}, Build:${buildNum} - ` +
+                      `Skipping preview processing because this PR did not touch any significant files.`);
+          res.sendStatus(204);
```
First `sendStatus()` then `log()`, for consistency with above :grin:

gkalpak commented on aio/aio-builds-setup/dockerbuild/scripts-js/test/upload-server/upload-server-factory.spec.ts:
```diff
@@ -134,15 +147,11 @@ describe('uploadServerFactory', () => {
+      buildCreator = UploadServerFactory.createBuildCreator(prs, defaultConfig.buildsDir, defaultConfig.domainName);
     });
 
-
```
I thought you liked these double empty lines 😕

petebacondarwin replied:
I'm fickle

gkalpak commented on aio/aio-builds-setup/dockerbuild/scripts-js/test/upload-server/upload-server-factory.spec.ts:
```diff
@@ -199,249 +208,221 @@ describe('uploadServerFactory', () => {
-          agent.patch('/health-check').expect(404),
-          agent.delete('/health-check').expect(404),
-        ], done);
+      it('should fail verifying the PR fails', async () => {
```
_fail_ --> _fail if_

gkalpak commented on aio/aio-builds-setup/dockerbuild/scripts-js/test/upload-server/upload-server-factory.spec.ts:
```diff
@@ -199,249 +208,221 @@ describe('uploadServerFactory', () => {
-          agent.get('/foo-health-check').expect(404),
-          agent.get('/foonhealth-check').expect(404),
-        ], done);
+      it('should fail creating the preview build fails', async () => {
```
_fail_ --> _fail if_

petebacondarwin replied:
cutnpaste

gkalpak replied:
I figured. So, I copied and pasted my comment too 😛

gkalpak reviewed (commented):
Reviewed everything except for `verify-setup/*`.
Great stuff 💯

<sub>(I soon gave up on my "alphabetic ordering everywhere" dream 😛)</sub>

Toxicable commented on .circleci/config.yml:
```diff
@@ -112,6 +112,22 @@ jobs:
+          path: *aio_preview_artifact_path
+          # The `destination` needs to be kept in synch with the value of
+          # `AIO_ARTIFACT_PATH` in `aio/aio-builds-setup/Dockerfile`
+          destination: aio/aio-snapshot.tgz
```
Is it possible to make the destination similar to the ones we have here?
https://github.com/petebacondarwin/angular/blob/6e9bc2bb6e44a76c56e3c7f1fb560c40c3127279/.circleci/config.yml#L92

If not then let me know and ill find out what kind of wiggle room we have with the size tracking website

petebacondarwin replied:
So what you are saying is that you want an extra forward slash in the destination?
We could go with `aio/dist/snaphsot.tgz` if you like?

Toxicable replied:
Yeah! that should work.
It sounds silly but from what I can tell it does require that 2nd flash for the frontend to process it correctly.

gkalpak replied:
But do we want to size-track this? (It contains images, guides, API docs, CSS styles, etc.)

petebacondarwin replied:
I think the idea is that the size tracking only happens for certain paths but that the system will break if there are not two slashes?

Toxicable replied:
Both correct.
I don't think tracking the size will bring much value here.
But right now it will track any and all artifacts, so the paths need to conform.
I can add in a path exclusion option to the config when i next get a chance

Toxicable replied:
Made an issue here: https://github.com/angular/github-robot/issues/12

petebacondarwin replied:
It would be good for the bot to be more lenient so that you don't have to track all PRs that might create artefacts.

Toxicable replied:
Fair point.
Maybe it should be a whitelist instead of a blacklist?

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/scripts-js/lib/verify-setup/jasmine-custom-matchers.ts:
```diff
@@ -0,0 +1,86 @@
+  public compare(actual: string, remove = true) {
+    const pass = checkFile(actual, remove);
+    return {
+      message: `Expected file at "${actual}" ${pass ? 'not' : ''} to exist`,
```
Doh!

gkalpak replied:
That's another problem I haven't found a satisfying solution to yet 😒
You have to compromize between source code readibility and output correctness :grin:

gkalpak replied:
<sub>
  One could use a tag function that removes double spaces 🤔
  I wonder if it is possible to overwrite the way untagged template literals are handled. Couldn't find anything.
</sub>

petebacondarwin replied:
perhaps we could put a backspace char (`\b`) instead of an empty string :-P

petebacondarwin replied:
that actually works!

```
> console.log('x \b y')
x y
```

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/scripts-js/lib/upload-server/build-verifier.ts:
```diff
@@ -1,87 +1,46 @@
-  // Constructor
-  constructor(protected secret: string, githubToken: string, protected repoSlug: string, organization: string,
+  /**
+   * Constructor a new BuildVerifier instance.
```
LOL!

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/scripts-js/lib/upload-server/build-retriever.ts:
```diff
@@ -0,0 +1,80 @@
+  if (!match) {
+    throw new Error(`No PR found in branch field: ${branch}`);
+  }
+  return +match![1];
```
That is what I thought too!

gkalpak replied:
Maybe we need to update TS. Latest TS definitely knows that stuff: [Proof][1]

[1]: http://www.typescriptlang.org/play/index.html#src=function%20test()%20%7B%0D%0A%20%20%20%20const%20match%20%3D%20%2Ffoo(123)%2F.exec('foo123')%3B%0D%0A%20%20%20%20const%20value1%20%3D%20match%5B1%5D%3B%20%20%2F%2F%20At%20this%20point%2C%20it%20could%20be%20%60null%60.%0D%0A%20%20%20%20if%20(!match)%20%7B%0D%0A%20%20%20%20%20%20%20%20throw%20new%20Error('null')%3B%0D%0A%20%20%20%20%7D%0D%0A%20%20%20%20const%20value2%20%3D%20match%5B1%5D%3B%20%20%2F%2F%20At%20this%20point%2C%20it%20couldn't%20be%20%60null%60.%0D%0A%7D%0D%0A

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/scripts-js/lib/upload-server/build-retriever.ts:
```diff
@@ -0,0 +1,80 @@
+
+  /**
+   * Get GitHub information about a build
+   * @param buildNum The number of the build for which for retrieve the info.
```
good job someone knows English around here

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/scripts-js/lib/upload-server/build-retriever.ts:
```diff
@@ -0,0 +1,80 @@
+export class BuildRetriever {
+  private logger = createLogger('BuildRetriever');
+  constructor(private api: CircleCIApi, private downloadSizeLimit: number, private downloadDir: string) {
+    assertNotMissingOrEmpty('downloadDir', downloadDir);
```
Yeah, because it is typed a `number` it can't be empty :-( So we could check that it is not `NaN` I guess or `>0` or something.

gkalpak replied:
I would argue that even `0` is an unwanted value (since it would disable the check altogether and is more likely to be a programming error than intentional).

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/scripts-js/lib/upload-server/index.ts:
```diff
@@ -1,34 +1,41 @@
+  create({
+    buildArtifactPath: AIO_ARTIFACT_PATH,
+    buildsDir: AIO_BUILDS_DIR,
+    circleCIToken: AIO_CIRCLECI_TOKEN,
```
`circleCiToken` it is then

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/scripts-js/test/common/github-api.spec.ts:
```diff
@@ -218,191 +183,162 @@ describe('GithubApi', () => {
-          mockResponse.statusCode = statusCode;
+    it('should add the \'Authorization\' header containing the \'githubToken\'', () => {
+      const requestHandler = nock('https://api.github.com')
+        .intercept(`/path`, 'method', undefined, {
```
Because it doesn't matter? Who says that backticked strings must have interpolation? The can also contain multiple lines with no interpolation.

gkalpak replied:
Some linter always barks at me when I use backticks without interpolation or newlines (or a tag function). If nothing else, for quoting consistency (e.g. "strings must be single-quoted").
It is probably ESLint :grin:
Anyway, I was just curious. If TSLint is happy, I am happy :smiley:

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/scripts-js/lib/clean-up/build-cleaner.ts:
```diff
@@ -3,29 +3,53 @@ import * as fs from 'fs';
-  // Methods - Protected
-  protected getExistingBuildNumbers(): Promise<number[]> {
-    return new Promise((resolve, reject) => {
+  public async cleanDownloads(openPrs: number[]) {
```
The previous approach of doing `(obj as any).protectedThing` caused me more than an hour of searching before I realised that a test was broken by a change to the API, since the TypeScript compiler wasn't able to spot the change.

So I took the brave decision to say that if it needed to be tested then it should be public :-)

gkalpak replied:
:+1:

Yeah, the `pricvate` + `(obj as any).xyz` approach is a pain indeed. (But I don't like making things `public` that shouldn't be either 😛)

I haven't found an ideal approach yet 😞
Nowadays, I try the following (far from ideal) approach:
- Make the methods `protected`.
- For tests, define a new class that extends the original one and makes the methods public.

alfaproject replied:
Shouldn't we only be testing public API surface? (or protected by overriding a class)

Private stuff is internal stuff that surely could be refactored at any point without breaking tests.

petebacondarwin replied:
Making a function protected just so that we can test it is as bad as making it public IMO.

I generally find that if you need to test a private thing that you can't get to, then that is an indication that the code needs restructuring - making that private thing a separately testable entity that can have a public interface.

So in this case, if we wanted to make [Martin Fowler](https://martinfowler.com/books/refactoring.html) happy, we should create something like this:

```typescript
interface Cleaner {
  clean(): Promise<void>;
}

class CleanerManager {
  constructor(private cleaners: Cleaner[]) {}
  public async cleanUp() { await Promise.all(this.cleaners.map(cleaner => cleaner.clean())); }
}

class BuildCleaner implements Cleaner {
  public async clean() {}
}
class DownloadCleaner implements Cleaner {
  public async clean() {}
}
```

Then we could test each of the "cleaners" independently but their implementation would be hidden to the client of the `CleanerManager`.

petebacondarwin replied:
But being a ["pragmatic programmer"](https://pragprog.com/) I think that is overkill in this case.
:-)

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/scripts-js/lib/common/github-pull-requests.ts:
```diff
@@ -1,46 +1,82 @@
+   * @returns A promise that is resolved with information about the requested PRs.
+   */
+  public fetchAll(state: PullRequestState = 'all') {
+    this.logger.log(`Fetching ${state} pull requests...`);
```
I guess the former. Shall I get rid of it?

gkalpak commented on aio/aio-builds-setup/dockerbuild/scripts-js/lib/clean-up/build-cleaner.ts:
```diff
@@ -3,29 +3,53 @@ import * as fs from 'fs';
+  public async cleanUp() {
+    try {
+      this.logger.log('Cleaning up builds and downloads');
+      const openPrs = await this.getOpenPrNumbers() || [];
```
Both (afaict) :grin:

gkalpak commented on aio/aio-builds-setup/dockerbuild/scripts-js/test/clean-up/build-cleaner.spec.ts:
```diff
@@ -1,143 +1,184 @@
-
+  beforeEach(() => {
+    spyOn(console, 'error');
+    spyOn(console, 'log');
```
It shouldn't be too difficult to change `createLogger()` to return a `Logger` instance (instead of an object literal) and spy on `Logger.prototype` instead. I.e.:

```ts
export class Logger {
  constructor(private scope: string) {}
  log(...args) { console.log(..., ...args); }
  ...
}

export function createLogger(scope: string) {
  return new Logger(scope);
}
```

Then in tests:
```diff
-spyOn(console, 'log');
+spyOn(Logger.prototype, 'log');
```

Although it would be a nice improvement (and a relatively easy change given that you have already done the hardest part: introducing a logger :heart:), it is not necessary for this PR (it is already much better than what we had before :grin:)

petebacondarwin commented:
@gkalpak  and @Toxicable please take another look

Toxicable reviewed (approved):
Size stuff LGTM

gkalpak reviewed (approved):
A couple of minor issues. LGTM otherwise :tada:

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/scripts-js/lib/common/circleci-api.ts:
```diff
@@ -3,7 +3,7 @@ import fetch from 'node-fetch';
 import {assertNotMissingOrEmpty} from './utils';
```
yeah... I knew that :-P

petebacondarwin replied:
Done

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/scripts-js/lib/upload-server/index.ts:
```diff
@@ -1,34 +1,41 @@
+    .create({
+      buildArtifactPath: AIO_ARTIFACT_PATH,
       buildsDir: AIO_BUILDS_DIR,
+      circleCIToken: AIO_CIRCLE_CI_TOKEN,
```
:-( will fix

petebacondarwin commented:
I've cleaned up after @gkalpak's comments.
Just needs approval from @IgorMinar now for `root` and `build-and-ci` groups.

petebacondarwin committed:
ci(docs-infra): add explicit return types to methods

petebacondarwin commented:
Rebased/squashed on current master and added a commit that adds return types to methods that should resolve https://github.com/angular/angular/pull/23576#discussion_r207682624

IgorMinar reviewed (approved):
In general this looks great. I left a few minor comments, which need to be resolved before this is merged. Thanks @petebacondarwin!

IgorMinar commented:
I see that there are two tasks that are still unresolved in the description:
- consider renaming the upload-server to preview-server - I'd very much like this to happen, because at this point there is no upload happening anywhere, so the name is quite confusing.
- how to run PWA-score test: this one is more tricky...

I see two options:
- an easy option would be to not run those tests against the preview servers and instead just run them as our regular monitoring tests 
- the other option would be to add a new a new job into [our CircleCI workflow](https://circleci.com/workflow-run/e57bcae9-f325-4db2-b7f3-044c3f7fbba2) that would run after we publish the build artifact and trigger the webhook (I assume that the webhook runs after each job). This new job could then poll the preview server until the preview is deployed (or timeout and fail), and only then run the tests.
- there is also the option that Pete mentioned in the description "by the preview-server triggering a specific CircleCI job via API" - this is a variation of first option I mentioned.

Do we have other options?

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/Dockerfile:
```diff
@@ -64,7 +73,7 @@ RUN mkdir /var/log/aio
 # Add extra package sources
 RUN apt-get update -y && apt-get install -y curl
-RUN curl --silent --show-error --location https://deb.nodesource.com/setup_6.x | bash -
+RUN curl --silent --show-error --location https://deb.nodesource.com/setup_8.x | bash -
```
It looks like the JS tests all pass so we should be fine with Node 10

petebacondarwin commented on aio/aio-builds-setup/dockerbuild/scripts-js/lib/verify-setup/nginx.e2e.ts:
```diff
@@ -252,31 +266,9 @@ describe(`nginx`, () => {
       });
 
 
-      it(`should reject files larger than ${h.uploadMaxSize}B (according to header)`, done => {
```
The files are no longer uploaded to the server. Instead the `BuildRetriever` downloads them.
There is a [test](https://github.com/angular/angular/pull/23576/files/8d311c112c5541b416b21972990669a8168b2f50#diff-a0a98001648888320b3fdf82b0e29f4aR111) that checks that we do not try to download files that are too large.

IgorMinar replied:
bah. of course! :)
thanks for the clarification

petebacondarwin commented:
Regarding the PWA test:

1) This is probably the simplest (and would be my preferred) option. The downside is that you cannot fail PRs if the tests fail. But in practice I don't believe that failed PWA is a strong reason to fail a PR anyway. Also these tests can often flake anyway due to network issues so the false negatives can often confused developers and add to the mental overhead of getting a PR ready. A regular test of the live sites should provide us with enough feedback to focus our efforts to fixing PWA failures when they have been merged to master.

2) I don't believe we can do this in a single workflow because (if I remember correctly) the artifact webhook is only triggered after the whole workflow completes.

3) Although this idea seems the most accurate, I don't think the additional complexity of setting this up provides much benefit over 1).

petebacondarwin committed:
ci(docs-infra): rename 'upload-server' to 'preview-server'

The server no longer has files uploaded to it. Instead it is more
accurate to refer to it as dealing with "previews" of PRs.

IgorMinar commented:
it seems that the last refactoring resulted in lint errors, @petebacondarwin can you please reformat the code? thanks!

petebacondarwin committed:
docs(docs-infra): the `build.sh` script was renamed to `create-image.sh`

mary-poppins commented:
You can preview fba0882 at https://pr23576-fba0882.ngbuilds.io/.
